L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,450 @@
+L46: C-core: New TLS Credentials API
+----
+* Author(s): ZhenLian
+* Approver: markdroth, yashykt, jiangtaoli2016
+* Status: Draft
+* Implemented in:
+* Last updated: September 9, 2020
+
+## Abstract
+
+This proposal aims to provide the following security features in a new TLS credential API:
+1. credential reloading: reload transport credentials periodically from various sources, e.g. from credential files, or users' customized Provider implementations, without shutting down
+2. custom authorization check: optionally bypass the certificate check, bypass the hostname check on the client side, and perform an additional per-connection authorization callback at the end of authentication
+3. TLS version configuration: optionally configure the minimum and maximum of the TLS version that will be used in the handshake
+
+## Background
+
+The current TLS implementation in gRPC C core has some restrictions:
+1. lack of the credential reloading support. Once an application starts, it can never reload its identity key/cert-chain and root certificates without shutting down
+2. inflexibility in the peer verification. In a mutual TLS scenario, servers would always perform certificate verification, and clients would always perform certificate verification as well as default hostname check. No customized checks can be introduced.
+3. inability to choose the TLS version that will be used
+
+There have always been some ad hoc requests on GitHub Issue Page for supporting these features. The demands for a more flexible and configurable API have increased with the emerging needs in cloud platform.
+One common example is to support SPIFFE ID as the identity format, which requires us to disable the default hostname check on the client side.
+Another example is to reload the identity certificates without shutting down the server, because certificates will eventually expire and the ability to rotate the certificates is essential to a cloud application.
+
+Hence we proposed an API that will meet the following requirements:  
+
+1. Flexible enough to support multiple use cases. Users should be able to use their own implementations if the provided ones don't fit their needs
+2. C core API needs to be simple, so that it can be easily wrapped by different languages
```

Comment:
Can we be more specific on it? It's not clear to me how "simple" the C-core API should be in order to get wrapped by other languages.

Reply:
That's a previous requirement. I think as time goes, this API becomes a bit more complicated, and some designs might be needed when wrapping the Core API. 
@jiangtaoli2016 
Would it be appropriate to remove this point here? I think our goal here is to make the API wrappable with other languages, but the simplicity is a bit vague and we have no way to measure it.

Reply:
I'd rephrase this as 
C core API needs to be simple enough, so that most of existing SSL credentials in wrapped languages can migrate to call this new API.